<img src="https://api.brandy.run/core/logo" width="80"/>

# CORE Code School

# Python Basics

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Variables-and-DataTypes" data-toc-modified-id="Variables-and-DataTypes-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Variables and DataTypes</a></span><ul class="toc-item"><li><span><a href="#Asigning-a-variable" data-toc-modified-id="Asigning-a-variable-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Asigning a variable</a></span></li><li><span><a href="#Checking-variable-types" data-toc-modified-id="Checking-variable-types-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Checking variable types</a></span></li><li><span><a href="#Casting" data-toc-modified-id="Casting-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Casting</a></span></li><li><span><a href="#Creating-a-new-list" data-toc-modified-id="Creating-a-new-list-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Creating a new list</a></span></li><li><span><a href="#Removing-elements" data-toc-modified-id="Removing-elements-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Removing elements</a></span></li><li><span><a href="#Index-and-Count" data-toc-modified-id="Index-and-Count-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Index and Count</a></span></li><li><span><a href="#Accessing-tuple-elements-and-f-strings" data-toc-modified-id="Accessing-tuple-elements-and-f-strings-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Accessing tuple elements and f-strings</a></span></li><li><span><a href="#Split,-Join-and-string-methods" data-toc-modified-id="Split,-Join-and-string-methods-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Split, Join and string methods</a></span></li><li><span><a href="#Counting-digits" data-toc-modified-id="Counting-digits-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>Counting digits</a></span></li></ul></li><li><span><a href="#Control-Flow" data-toc-modified-id="Control-Flow-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Control Flow</a></span><ul class="toc-item"><li><span><a href="#Guided-example" data-toc-modified-id="Guided-example-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Guided example</a></span></li><li><span><a href="#Simple-calculator" data-toc-modified-id="Simple-calculator-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Simple calculator</a></span></li><li><span><a href="#Dog-Registry" data-toc-modified-id="Dog-Registry-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Dog Registry</a></span></li><li><span><a href="#Largest-Prime" data-toc-modified-id="Largest-Prime-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Largest Prime</a></span></li></ul></li></ul></div>

# ¿Que es un lenguaje de programación?

In [ ]:
from IPython.display import IFrame
IFrame('https://www.youtube.com/embed/XASY30EfGAc"', width=560, height=315)

Fuente: Women Techmakers

## Variables and DataTypes

Los ejercícios de la sección 1 acompañan la lección `Variables and Data Types` de el prework. Practicaremos con los tipos basicos de datos en Python.

### Asigning a variable

En la celda abajo, asigne un valor entre 0 y 100 a la variable x, imprimela, asignale un valor que sea un texto y imprimela otra vez. Sustituye las lineas de los comentarios.

In [ ]:
x = 100
print(type(x))
x = "Hello"
print(type(x))
x = 19.4
print(type(x))
x = True
print(type(x))
x = [1,2,3,4]
print(x)
x.append(5)
print(x)
print(type(x))
alumno = {
    "nombre":"Pepe",
    "edad":10,
    "b":2
}
print(type(x))
x = (1,2,3,4)
print(type(x))
x = set([1,2,3,4])
print(type(x))
x = None
print(type(x))
#x = nan
#print(type(x))
x = lambda x:3
print(type(x))

class A:
  def __init__(self):
    self.x = 1

print(type(A))

print(type(Exception("1234")))

print(type(b'123'))


<class 'int'>
<class 'str'>
<class 'float'>
<class 'bool'>
[1, 2, 3, 4]
[1, 2, 3, 4, 5]
<class 'list'>
<class 'list'>
<class 'tuple'>
<class 'set'>
<class 'NoneType'>
<class 'function'>
<class 'type'>
<class 'Exception'>
<class 'bytes'>


In [ ]:
print(set([1,2,3,4,5,5,5,5,5,5,5,5]))

{1, 2, 3, 4, 5}


### Checking variable types
Asigna diferentes tipos de datos a variables y prueba el resultado de la función `type`. Los 4 tipos escritos son sugerencias, pero puedes probar los que quieras. 

**Bonus:** Escribe un condicional `if` para de una lista de valores imprimir solo aquellos que sean de un tipo específico.

In [ ]:
entero = 1
decimal = 0.1
lista = [entero, decimal]
dicionario = {'nombre': 'alejandro'}

for element in entero, decimal, lista, dicionario:
  if isinstance(element, int):
    print(type(element))


<class 'int'>


In [ ]:
x = []
isinstance(x,list)

True

### Casting
Escribe un programa en python que itere sobre una lista de valores de diferentes tipos y les sume todos.

In [ ]:
lista = [1,2,"3.0",4.5, "6", 9, 10, "42"]

sum = 0
for element in lista:
  sum += float(element)

print(sum)

#print(sum(map(float, lista)))

77.5


In [ ]:
x = 5
print(type(x),x)
y = str(x)
print(type(y),y)
print(x*2)
print(y*2)

<class 'int'> 5
<class 'str'> 5
10
55


In [ ]:
x = "60a"
print(x, type(x))
y = int(x)
print(y, type(y))

60a <class 'str'>


ValueError: ignored

### Creating a new list
Defina en una variable una lista vacia. Eso se hace con apenas corchetes, i.e.: `x=[]`

En seguida, utilizando un bucle for, itere por los numeros del 1 al 200 y añada a la lista solo los multiplos de 3. Cuente cuantos elementos hay en la lista final.

**TIP:** Para saber la longitud de una lista se puede usar la función `len`: len(lst)

In [ ]:
x = []

for i in range(1,200):
  if i % 3 == 0:
    x.append(i)

print(x)
print(len(x))

# x=[ i for i in range(1, 200) if i%3==0]
# print(f"Len: {len(x)}")


[3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99, 102, 105, 108, 111, 114, 117, 120, 123, 126, 129, 132, 135, 138, 141, 144, 147, 150, 153, 156, 159, 162, 165, 168, 171, 174, 177, 180, 183, 186, 189, 192, 195, 198]
66


In [ ]:
x = []
# Step is 3
for i in range(3,200,3):
  x.append(i)
print(len(x))

66


### Removing elements

De la siguiente lista, remova todos los numeros negativos. Imprima la lista resultante.

**Bonus:** Imprime la lista resultante ordenada ascendientemente y hazlo alterando la lista existente, sin crear una nueva.

In [ ]:
lst = [-125, 243, 332, -70, 310, 213, -349, 499, 164, 74, 327, 87, -298, -281, 184, 212.5, -97, 22, -9, -246, 392, 
       -397, 471, -354, 195, 359, 136, -100, -350, -50, 329, -331, 218, -418, 415, 469.77, -182, -129, -328, -450, 
       293, -485, -333, 427, -291, -413, 0, 157, 76, 81, 35, 345, 337, 20, 94.21, 65, -120, 185, -192, 120, 491, 
       453, -472, 122.11, -242, 416, 86, 114, 229, -312, 228, -284, 122, -256, -326, 425, 252, 404, -439, -457, -298, 
       297, -133, 490, -454, -193.32, 235, -460, 422, 379, -438, -291, 344, 41.5, 310.0, -199, 113, 484, 437, 47]

x = []
for e in lst:
  if e>=0:
    x.append(e)

print(x)


lst = [item for item in lst if item >= 0] # list comprehension
lst.sort()

# sorted(creates a new list...)
#lst = sorted(list(filter(lambda x: x >= 0, lst)))

print(lst)

[243, 332, 310, 213, 499, 164, 74, 327, 87, 184, 212.5, 22, 392, 471, 195, 359, 136, 329, 218, 415, 469.77, 293, 427, 0, 157, 76, 81, 35, 345, 337, 20, 94.21, 65, 185, 120, 491, 453, 122.11, 416, 86, 114, 229, 228, 122, 425, 252, 404, 297, 490, 235, 422, 379, 344, 41.5, 310.0, 113, 484, 437, 47]
[0, 20, 22, 35, 41.5, 47, 65, 74, 76, 81, 86, 87, 94.21, 113, 114, 120, 122, 122.11, 136, 157, 164, 184, 185, 195, 212.5, 213, 218, 228, 229, 235, 243, 252, 293, 297, 310, 310.0, 327, 329, 332, 337, 344, 345, 359, 379, 392, 404, 415, 416, 422, 425, 427, 437, 453, 469.77, 471, 484, 490, 491, 499]


### Index and Count

A partir de la siguiente lista:
- Cuenta cuantas veces sale el numero 0
- Crea una list con los indices de todas las ocurrencias del numero 1

In [ ]:
lst = [4, 6, 1, 8, 6, 4, 3, 5, 5, 9, 7, 2, 3, 6, 4, 0, 8, 5, 6, 3, 9, 3, 7, 1, 7, 5, 7, 4, 6, 1, 9, 0, 6, 6, 3, 7, 
       2, 1, 2, 6, 7, 2, 1, 7, 9, 7, 1, 3, 6, 5, 6, 4, 6, 5, 1, 2, 0, 5, 9, 9, 8, 3, 7, 3, 4, 6, 1, 5, 5, 4, 1, 9, 
       3, 0, 5, 5, 1, 5, 1, 8, 4, 5, 6, 8, 4, 2, 2, 4, 0, 7, 6, 9, 3, 0, 3, 2, 8, 5, 1, 8]

sum = 0
idx = []

for index, item in enumerate(lst):
  if item == 0:
    sum += 1
  if item == 1:
    idx.append(index)

print(sum)
print(idx)

# print(lst.count(0))

# l = [ idx for idx, i in enumerate(lst) if i == 1 ]
# print(l)

6
[2, 23, 29, 37, 42, 46, 54, 66, 70, 76, 78, 98]


In [ ]:
x = ["e","a","u","q"]
i = 0
for e in x:
  print("Index",i, "element",e)
  i = i+1


Index 0 element e
Index 1 element a
Index 2 element u
Index 3 element q


In [ ]:
x = ["e","a","u","q"]

for i,e in enumerate(x):
  print("Index",i, "element",e)
  


In [ ]:
x = ["e","a","u","q"]

list(enumerate(x))

[(0, 'e'), (1, 'a'), (2, 'u'), (3, 'q')]

In [ ]:
x = ["ab","cd","ef","gh"]
for m,n in x:
  print(m,n) 

a b
c d
e f
g h


In [ ]:
for e in "holaquetal":
  print("letra:",e)

letra: h
letra: o
letra: l
letra: a
letra: q
letra: u
letra: e
letra: t
letra: a
letra: l


In [ ]:
len(set("hola que tal"))

9

In [ ]:
set("hola que tal")

{' ', 'a', 'e', 'h', 'l', 'o', 'q', 't', 'u'}

### Accessing tuple elements and f-strings

En la siguiente celta se encuentran dos puntos, llamados `start` y `finish`, con formato `(horas, minutos, segundos)`.
Calcule:

- Cuantos segundos pasaron entre start y finish
- La misma respuesta en el formato `"Total: X horas, Y minutos y Z segundos"`

In [ ]:
start = (2,25,13)
finish = (18,3,55)

from datetime import datetime

start_string = ' '.join(map(str, start))
finish_string = ' '.join(map(str, finish))

start_obj = datetime.strptime(start_string, "%H %M %S")
finish_obj = datetime.strptime(finish_string, "%H %M %S")

diff = finish_obj - start_obj

print(diff.total_seconds())

hours = int(diff.total_seconds() / 3600)
minutes = int(diff.total_seconds() / 60 % 60)
seconds = int(diff.total_seconds() % 60)

print(f'Total: {hours} horas, {minutes} minutos y {seconds} segundos')


56322.0
Total: 15 horas, 38 minutos y 42 segundos


In [ ]:
start = (2,25,13)
finish = (18,3,55)

h,m,s = start
total_start = s + m*60 + h*60*60

h,m,s = finish
total_end = s + m*60 + h*60*60
print(total_end-total_start)

56322


### Split, Join and string methods

Usando la siguiente cita del programador Martin Fowler:
- Imprime la cita al contrario
- Genere una lista de todas las palabras presentes en orden alfabetico.
- Una todas las palabras de la lista anterior con el caracter `~` como separador. Ejemplo:`"a~any~can~can~c..."`

In [ ]:
quote = "Any fool can write code that a computer can understand. Good programmers write code that humans can understand."

print(quote[::-1])

print(sorted(quote.lower().split()))
# print(sorted(q, key=str.lower))

print("~".join(quote.split()))

.dnatsrednu nac snamuh taht edoc etirw sremmargorp dooG .dnatsrednu nac retupmoc a taht edoc etirw nac loof ynA
['a', 'any', 'can', 'can', 'can', 'code', 'code', 'computer', 'fool', 'good', 'humans', 'programmers', 'that', 'that', 'understand.', 'understand.', 'write', 'write']
Any~fool~can~write~code~that~a~computer~can~understand.~Good~programmers~write~code~that~humans~can~understand.


### Counting digits
En ese ultimo ejercício de la sección 1, crea un bucle for iterando por los numeros de 0 a 100. Deberás:
- Contar cuantas veces aparece cada digito del 0 al 9 en todos los numeros de ese rango. El resultado esperado es un diccionario cuyas keys sean `1,2,3,4,5,6,7,8,9,0` y los values el numero de repeticiones de cada uno.
- Crear una lista de los numeros cuyos digitos no se repiten. Deberan incluirse todos los numeros 1 digito y numeros como el 102, pero no 101.

In [ ]:
dictionary = {
    '0': 0,
    '1': 0,
    '2': 0,
    '3': 0,
    '4': 0,
    '5': 0,
    '6': 0,
    '7': 0,
    '8': 0,
    '9': 0,
}

uniques = []

for x in range(0,100):
  lst = [str(a) for a in str(x)]
  if len(lst) == len(set(lst)):
    uniques.append(int(''.join(lst)))
  for digit in lst:
    if digit in dictionary:
      dictionary[digit] += 1

print(dictionary)
print(uniques)

{'0': 10, '1': 20, '2': 20, '3': 20, '4': 20, '5': 20, '6': 20, '7': 20, '8': 20, '9': 20}
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98]


## Control Flow
Los ejercícios de la sección 2 acompañan la lección `Control Flow` de el prework. Aplicaremos los conceptos de la sección anterior y praticaremos la organización de datos y procesos.

### Guided example
Encontrar el mayor numero palíndromo en una lista de 1000 numeros aleatórios entre 10 y 10000.

In [ ]:
import numpy as np
# Generamos una lista de numeros aleatorios
nums = np.random.randint(10,10_000, 1_000)

- Abordaje 1

Separar los palíndromos primero y buscar el mayor dentre ellos

In [ ]:
%%timeit -n 1000 # Esta linea haz con que jupyter calcule el tiempo medio de ejecución de esa celda en mil repeticiones

palindromes = []
for num in nums:
    num = str(num)
    if num == num[::-1]:
        palindromes.append(int(num))
        
max_palindrome = max(palindromes)

1000 loops, best of 5: 906 µs per loop


- Abordaje 2

Ordenar primero todos los numeros del mayor al menor, despues buscar los palíndromos y parar cuando encontremos uno.

In [ ]:
%%timeit -n 1000 # Esta linea haz con que jupyter calcule el tiempo medio de ejecución de esa celda en mil repeticiones

nums_sorted = sorted(nums, reverse=True)

for num in nums:
    num = str(num)
    if num == num[::-1]:
        max_palindrome = num
        break

1000 loops, best of 5: 398 µs per loop


Utilizando de las herramientas de control y organización de los datos, resuelva los siguientes ejercícios:

### Simple calculator
En ese ejercício debes escribir un programa que funcione como una calculadora. Los valores `a` y `b` deben ser numeros enteros y `operation` debe ser uno de los caracteres `'+', '-', '*', '/'`.
Despues que el usuario escribir los 3 argumentos, la calculadora debe imprimir el resultado de la operación. 

**TIP:** La función `input` siempre devuelve como resultado un `str`, tendrás que hacer casting al tipo correcto.

In [ ]:
a = int(input("Type in first number: "))
operation = input("What is the desired operation? ")
b = int(input("Type in the second number: "))

def calculator(a, operation, b):
  if operation == '+':
    return a+b
  elif operation == '-':
    return a-b
  elif operation == '*':
    return a*b
  elif operation == '/':
    return a/b

calculator(a, operation, b)

Type in first number: 2
What is the desired operation? +
Type in the second number: 2
Type in first number: 2
What is the desired operation? +
Type in the second number: 2


KeyboardInterrupt: ignored

### Dog Registry
Un amigo que es cuidador de perros ha apuntado datos sobre los perros que cuida. Como se enteró de que estás haciendo un bootcamp de Data, él te ha pedido ayuda para sacar algunas informaciones de esos datos:
- ¿Quien es el dueño al cual cuida los perros más días en la semana?
- ¿Cual es el día de la semana en que cuida de más perros?
- ¿Quien es el dueño que tiene más perros?

**Bonus:** ¿Harías algún cambio a la manera como estan organizados los datos?

In [ ]:
data = [('Oreo', 'Liam', 'Thursday'),
        ('Chippy', 'Olivia', 'Tuesday'),
        ('Luna', 'Olivia', 'Friday'),
        ('Cujo', 'Noah', 'Friday'),
        ('Cooper', 'Emma', 'Friday'),
        ('Rex', 'Emma', 'Friday'),
        ('Daisy', 'Emma', 'Friday'),
        ('Molly', 'Oliver', 'Thursday'),
        ('Buddy', 'Oliver', 'Tuesday'),
        ('Tucker', 'Ava', 'Monday'),
        ('Bentley', 'Elijah', 'Monday'),
        ('Riley', 'Charlotte', 'Monday'),
        ('Bear', 'William', 'Wednesday'),
        ('Duke', 'Sophia', 'Tuesday'),
        ('Jax', 'James', 'Sunday'),
        ('Rocky', 'Amelia', 'Thursday'),
        ('Lily', 'Benjamin', 'Saturday'),
        ('Milo', 'Isabella', 'Sunday'),
        ('Piper', 'Lucas', 'Monday'),
        ('Toby', 'Lucas', 'Monday'),
        ('Zeus', 'Lucas', 'Monday'),
        ('Ellie', 'Mia', 'Friday'),
        ('Nala', 'Mia', 'Wednesday'),
        ('Roxy', 'Mia', 'Monday'),
        ('Winston', 'Mia', 'Tuesday'),
        ('Gracie', 'Henry', 'Wednesday'),
        ('Coco', 'Evelyn', 'Monday'),
        ('Murphy', 'Alexander', 'Sunday'),
        ('Dexter', 'Harper', 'Monday')]

max_day=max(set([item[2] for item in data]), key=[item[2] for item in data].count)
max_owner=max(set([item[1] for item in data]), key=[item[1] for item in data].count)

print(f'His busier day: {max_day}')
print(f'The most happy owner: {max_owner}')
print(f'The owner who gives him more happiness is {max_owner}')

# I'd don't change the data structure in any way, but improve my skills to be able
# to manage it in an easier way.

# But if it's about how data is organized strictly speaking, I'd put owners in the
# first row

His busier day: Monday
The most happy owner: Mia
The owner who gives him more happiness is Mia


### Largest Prime

Escribe un programa en que dado un valor `n`, encuentre el mayor numero primo menor que `n`.

**TIP:** Un numero primo es el que no es divisible por ningun valor salvo 1 y si mismo.

In [ ]:
n = int(input("Input maximum value: "))

prime = False
for number in reversed(range(n)):
  if prime:
    break
  else:
    for i in range(2, number-1):
      if number % i == 0:
        break
      else:
        prime = True
        print(number)
        break
        


Input maximum value: 9
Inner 7
